### PR TITLE
feat: Publish/load pre-quantized models

### LINKED ISSUE #4: Save the 4-bit quantized version of the Llama2 7b model in blob storage
We currently load the model from the blob storage in the ML workspace: https://autoraml3241530052.blob.core.windows.net/azureml-blobstore-b7ef477b-ca4a-44e3-a029-0e0542bdcd47/base_models/llama-2-7b-chat-hf/

Using [this code](https://github.com/AutoResearch/autodoc/blob/dea9d8ce77f624e5088384759739f3e073c62980/src/autora/doc/runtime/predict_hf.py#L14), which loads the full model and quantize it during load time. This takes about 6 mins.

If we can save a copy of the model with a new name, e.g. `llama-2-7b-chat-hf-4bit`, the loading time should significantly decrease.

The easiest way to do this is probably in a notebook that runs in the workspace and then something along these lines:

```py
from azureml.core import Workspace, Datastore

new_name = "llama-2-7b-chat-hf-4bit"
model = ... # load and quantize model like in the code predict_hf.py code above

# save it locally
model.save_pretrained(f"./models/{new_name}")

# double check that the quantized model can be loaded too ...

# If all goes well, upload to blob storage:
workspace = Workspace.from_config()
ds = workspace.get_default_datastore()
ds.upload(f"./models/{new_name}", f"./base_models/{new_name}", show_progress=True, overwrite=True)


# verify the model can be loaded from blob storage by submitting a new prediction job with the new model. See README.md
```

### LINKED ISSUE #8: Support model downloading/caching
The models are too big to ship with the package, so a better approach is to download them on first use and cache them locally.

Subtasks:
- Create an org in Huggingface
- Upload the models to the org (two versions? 7b and 7b quantized?
- Create a mapping of a model version+platform to physical model file
- Add logic to download on-demand
- Support model versioning and checking for newer models

### PR DESCRIPTION
**Pull request by carlosgjs**
Running quantized models significantly reduces the GPU memory required for inference. Instead of downloading the full model and quantizing it during the load, we can quantize the model offline and save it. At runtime the (smaller) quantized model can be loaded.

This PR includes 3 changes:
- Updates to the `bitsandbytes` and `transformers` versions which support quantizing
- A notebook for quantizing and publishing models
- Logic to map known models to their pre-quantized versions.

Closes #4 
Closes #8

### CODE REVIEW

File: notebooks/import_model.ipynb [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    "}"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "tokenizer = AutoTokenizer.from_pretrained(model_path)\n",
```
Comment (uwcdc) [inline notebooks/import_model.ipynb:52]: I would add a comment here for more clarity about what this cell does.

File: notebooks/import_model.ipynb [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# model_path = \"codellama/CodeLlama-7b-Instruct-hf\"\n",
+    "# model_path = \"meta-llama/Llama-2-7b-chat-hf\"\n",
```
Comment (uwcdc) [inline notebooks/import_model.ipynb:22]: Should these lines be here? If so, what are they for?

**Review by uwcdc — COMMENTED** [review]
Those were the only 2 things that I had questions about. But I love the updated formatting in the other files.

File: notebooks/import_model.ipynb (hunk 2 above)
Comment (carlosgjs) [reply notebooks/import_model.ipynb]: removed

File: notebooks/import_model.ipynb (hunk 1 above)
Comment (carlosgjs) [reply notebooks/import_model.ipynb:48]: done

File: notebooks/generate.ipynb [hunk 3]
```diff
@@ -25,12 +25,9 @@
     "# Uncomment to clone and install autodoc from GitHub\n",
     "# !pip uninstall -y autora-doc\n",
     "# !git clone https://github.com/AutoResearch/autodoc.git\n",
-    "# !pip install -e \"./autodoc[cuda,train]\"\n",
+    "# !pip install \"./autodoc[cuda,train]\"\n",
     "\n",
-    "# Login to Huggingface since access to the model repo is private\n",
-    "# 1) Request access through: https://ai.meta.com/resources/models-and-libraries/llama-downloads/\n",
-    "# 2) Get a Huggingface token from: https://huggingface.co/settings/token (use same email as above)\n",
-    "# !huggingface-cli login --token <your HF token>"
```
Comment (uwcdc) [inline notebooks/generate.ipynb:33]: Do you want the user to hard code the HF token into the notebook or pull it from an env variable?

File: notebooks/import_model.ipynb [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    "model = AutoModelForCausalLM.from_pretrained(model_path, quantization_config=conf, device_map=\"auto\")"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "# Running this requires loging into HuggingFace via `huggingface-cli login --token` using a write token\n",
```
Comment (uwcdc) [inline notebooks/import_model.ipynb:58]: typo: logging

File: src/autora/doc/runtime/predict_hf.py [hunk 5]
```diff
@@ -84,19 +88,24 @@ def tokenize(self, input: List[str]) -> Dict[str, List[List[int]]]:
...
-            return {
-                "quantization_config": BitsAndBytesConfig(
-                    load_in_4bit=True,
-                    bnb_4bit_use_double_quant=True,
-                    bnb_4bit_quant_type="nf4",
-                    bnb_4bit_compute_dtype=torch.bfloat16,
-                ),
-                "device_map": "auto",
-            }
+            config["quantization_config"] = BitsAndBytesConfig(
```
Comment (uwcdc) [inline src/autora/doc/runtime/predict_hf.py:103]: I like this change a LOT. Makes the returns more consistent and overall code more readable.

**Review by uwcdc — COMMENTED** [review]
Overall good. Just a couple small fixes + a compliment.

File: notebooks/generate.ipynb (hunk 3 above)
Comment (carlosgjs) [reply notebooks/generate.ipynb:33]: This was removed, no login required since we switched to a public model

**Review by uwcdc — APPROVED** [review]
LGTM!

**Pull request merged**